File: 프로그래머스/lv2/72411. 메뉴 리뉴얼/메뉴 리뉴얼.py
from itertools import combinations
from collections import Counter

def solution(orders, course):
    res = []
    orders = [set(o) for o in orders]
    
    for i in course:
        c_n = Counter() 
        for a,b in combinations(orders,2):
            for c in combinations(a&b,i):
                c_n += Counter([''.join(sorted(c))])

        if m_v := max(c_n.values(),default=0):
            res += [k for k,v in c_n.items() if v==m_v]
    
    return sorted(res)

File: 프로그래머스/lv2/72411. 메뉴 리뉴얼/메뉴 리뉴얼.js
function* combination(arr, n) {
  if (n == 1) {
    yield* arr.map((v) => [v]);
    return;
  }
  for (let i = 0; i < arr.length; i++) {
    for (let rest of combination(arr.slice(i + 1), n - 1)) {
      yield [arr[i], ...rest];
    }
  }
}

function solution(orders, course) {
  res = [];
  for (let n of course) {
    sets = {}
    for (let [a, b] of combination(orders, 2)) {
      set = a.split("").filter((v) => b.split("").includes(v)).sort();
      for (let candiate of combination(set, n)) {
        key = candiate.join("");
        if (!sets[key]) {
          sets[key] = 1;
        } else {
          sets[key] += 1;
        }
      }
    }

    _max = Object.entries(sets).reduce((acc,cur) => {
      if (cur[1] > acc[1]) {
        return [[cur[0]], cur[1]]
      } else if (cur[1] == acc[1]) {
        return [[cur[0], ...acc[0]], acc[1]]
      }
      return acc
    }, [[],-Infinity])
    res = res.concat(_max[0])
  }
  return res.sort()
}